lib/firmware/cosmos.c

```diff
@@ -0,0 +1,107 @@
+#include "keepkey/firmware/cosmos.h"
+
+#include "keepkey/board/confirm_sm.h"
+#include "keepkey/board/util.h"
+#include "keepkey/firmware/home_sm.h"
+#include "keepkey/firmware/storage.h"
+#include "trezor/crypto/segwit_addr.h"
+#include "trezor/crypto/ecdsa.h"
+#include <stdbool.h>
+#include <time.h>
+
+bool cosmos_path_mismatched(const CoinType *_coin,
+                            const uint32_t *address_n,
+                            const uint32_t address_n_count)
+{
+    // m/44' : BIP44-like path
+    // m / purpose' / bip44_account_path' / account' / x / y
+    bool mismatch = false;
+    mismatch |= address_n_count != 5;
+    mismatch |= address_n_count > 0 && (address_n[0] != (0x80000000 + 44));
+    mismatch |= address_n_count > 1 && (address_n[1] != _coin->bip44_account_path);
+    mismatch |= address_n_count > 2 && (address_n[2] & 0x80000000) == 0;
+    mismatch |= address_n_count > 2 && (address_n[3] & 0x80000000) == 0;
```

Comment:
undefined behavior
```suggestion
    mismatch |= address_n_count > 3 && (address_n[3] & 0x80000000) == 0;
```

Reply:
also, x and y shouldn't be hardened. maybe you meant `!= 0`?

Reply:
or are they supposed to be zero, ie:
```
mismatch |= address_n_count > 3 && address_n[3] != 0;
```

Reply:
great catch, no they need to not be hardened.